### DISCUSSION TITLE
How to connect ferretdb with Java program

### DISCUSSION THREAD

**Discussion by gaoyan0204** [FerretDB/FerretDB#991] [General]
how to set the user and password of mongodb？
In addition Can I connect ferretdb through Java program in the way of mongodb? Now the program reports an error.
com.mongodb.MongoSocketReadException: Prematurely reached end of stream。

**Comment by AlekSi** [comment]
The username and password should be empty.

May you provide a FerretDB output when you see that error?

**Reply by gaoyan0204** [reply]
ferretdb  log：

2022-08-05T00:17:19.208Z	ERROR	// 192.168.200.179:63913 -> 192.168.80.3:27017 	clientconn/conn.go:352	Handler error for unhandled response opcode{error 26 0  NotImplemented (238): CmdQuery: unhandled command "buildinfo"} {opcode 25 0  OP_REPLY}
ferretdb        | github.com/FerretDB/FerretDB/internal/clientconn.(*conn).route
ferretdb        | 	github.com/FerretDB/FerretDB/internal/clientconn/conn.go:352
ferretdb        | github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run
ferretdb        | 	github.com/FerretDB/FerretDB/internal/clientconn/conn.go:176
ferretdb        | github.com/FerretDB/FerretDB/internal/clientconn.(*Listener).Run.func2
ferretdb        | 	github.com/FerretDB/FerretDB/internal/clientconn/listener.go:149
ferretdb        | 2022-08-05T00:17:19.208Z	DEBUG	// 192.168.200.179:63913 -> 192.168.80.3:27017 	clientconn/conn.go:436	Response header: length:     0, id:    0, response_to:    0, opcode: OP_REPLY
ferretdb        | 2022-08-05T00:17:19.208Z	DEBUG	// 192.168.200.179:63913 -> 192.168.80.3:27017 	clientconn/conn.go:437	Response message:
ferretdb        | <nil>



ferretdb        | 2022-08-05T00:17:19.208Z	DPANIC	// 192.168.200.179:63913 -> 192.168.80.3:27017 	clientconn/conn.go:130	runtime error: invalid memory address or nil pointer dereference
ferretdb        | (err = <nil>)
ferretdb        | github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run.func2
ferretdb        | 	github.com/FerretDB/FerretDB/internal/clientconn/conn.go:130
ferretdb        | runtime.gopanic
ferretdb        | 	runtime/panic.go:844
ferretdb        | runtime.panicmem
ferretdb        | 	runtime/panic.go:220
ferretdb        | runtime.sigpanic
ferretdb        | 	runtime/signal_unix.go:818
ferretdb        | github.com/FerretDB/FerretDB/internal/wire.(*OpReply).MarshalBinary
ferretdb        | 	github.com/FerretDB/FerretDB/internal/wire/op_reply.go:92
ferretdb        | github.com/FerretDB/FerretDB/internal/wire.WriteMessage
ferretdb        | 	github.com/FerretDB/FerretDB/internal/wire/msg_body.go:100
ferretdb        | github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run
ferretdb        | 	github.com/FerretDB/FerretDB/internal/clientconn/conn.go:236
ferretdb        | github.com/FerretDB/FerretDB/internal/clientconn.(*Listener).Run.func2
ferretdb        | 	github.com/FerretDB/FerretDB/internal/clientconn/listener.go:149
ferretdb        | panic: runtime error: invalid memory address or nil pointer dereference [recovered]
ferretdb        | 	panic: runtime error: invalid memory address or nil pointer dereference
ferretdb        | (err = <nil>)
ferretdb        | [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xb76c77]
ferretdb        | 
ferretdb        | goroutine 29 [running]:
ferretdb        | go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc0000e2600, {0x0, 0x0, 0x0})
ferretdb        | 	go.uber.org/zap@v1.21.0/zapcore/entry.go:232 +0x60a
ferretdb        | go.uber.org/zap.(*SugaredLogger).log(0xc000010240, 0x3, {0x143513a, 0xd}, {0xc000161200, 0x2, 0x2}, {0x0, 0x0, 0x0})
ferretdb        | 	go.uber.org/zap@v1.21.0/sugar.go:227 +0x13b
ferretdb        | go.uber.org/zap.(*SugaredLogger).DPanicf(...)
ferretdb        | 	go.uber.org/zap@v1.21.0/sugar.go:154
ferretdb        | github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run.func2()
ferretdb        | 	github.com/FerretDB/FerretDB/internal/clientconn/conn.go:130 +0x1ab
ferretdb        | panic({0x13097e0, 0x1c5a1f0})
ferretdb        | 	runtime/panic.go:844 +0x258
ferretdb        | github.com/FerretDB/FerretDB/internal/wire.(*OpReply).MarshalBinary(0x0)
ferretdb        | 	github.com/FerretDB/FerretDB/internal/wire/op_reply.go:92 +0x57
ferretdb        | github.com/FerretDB/FerretDB/internal/wire.WriteMessage(0xc0001380f0?, 0xc000134330, {0x15fb310, 0x0})
ferretdb        | 	github.com/FerretDB/FerretDB/internal/wire/msg_body.go:100 +0x7b
ferretdb        | github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run(0xc0001380f0, {0x15fa8e8?, 0xc0001312c0})
ferretdb        | 	github.com/FerretDB/FerretDB/internal/clientconn/conn.go:236 +0x1ce5
ferretdb        | github.com/FerretDB/FerretDB/internal/clientconn.(*Listener).Run.func2()
ferretdb        | 	github.com/FerretDB/FerretDB/internal/clientconn/listener.go:149 +0xd25
ferretdb        | created by github.com/FerretDB/FerretDB/internal/clientconn.(*Listener).Run
ferretdb        | 	github.com/FerretDB/FerretDB/internal/clientconn/listener.go:107 +0x785

**Reply by AlekSi** [reply]
What MongoDB driver you use and what version? If not the latest, may you update it?

**Reply by gaoyan0204** [reply]
<dependency>
			<groupId>org.mongodb</groupId>
			<artifactId>mongo-java-driver</artifactId>
			<version>3.12.10</version>
		</dependency>

Is there a problem with this version? Which version should I use？

**Reply by gaoyan0204** [reply]
<dependency>
			<groupId>org.mongodb</groupId>
			<artifactId>mongodb-driver-sync</artifactId>
			<version>4.7.0</version>
		</dependency>

I'll change the driver to this. The problem still exists.The mistakes are the same.

**Reply by AlekSi** [reply]
The problem is that OP_QUERY message type is deprecated since MongoDB 5.0, so FerretDB does not implement it. The latest version of the driver should not send `buildinfo` command in OP_QUERY message. May you provide a small self-contained example with build instructions so we could take a look?
